source/extensions/quic_listeners/quiche/udp_gso_batch_writer.cc

```diff
@@ -0,0 +1,129 @@
+#include "extensions/quic_listeners/quiche/udp_gso_batch_writer.h"
+
+#include "common/network/io_socket_error_impl.h"
+
+#include "extensions/quic_listeners/quiche/envoy_quic_utils.h"
+
+namespace Envoy {
+namespace Quic {
+namespace {
+Api::IoCallUint64Result convertQuicWriteResult(quic::WriteResult quic_result) {
```

Comment:
I think there is some inconsistency between the result conversion here and UdpListenerImpl::send().
The caller may need the IoCallUint64Result to decide what do to next, i.e. how much data to drain from the buffer.
For WRITE_STATUS_OK and WRITE_STATUS_BLOCKED_DATA_BUFFERED, the caller wouldn't need to resend the data again. So it should be able to drain the whole packet buffer. Currently it drains rc_ bytes of data. If we return 0 in this case, the buffer will remain undrained. And at a certain point when another write call triggers flush, the rc_ is larger than the packet the caller is trying to send.

I think we have two options here:
1. change the callers to decide how much data to drain based on returned IoError object. If it's nullptr, that means data is either buffered in the writer or sent on the wire. The whole packet can be drained.
2. instead of return the actual rc_, return the size of the packet passed into writePacket() upon quic::WRITE_STATUS_OK and quic::WRITE_STATUS_BLOCKED_DATA_BUFFERED

Reply:
This was one of the topics I wanted to discuss with you. I have made changes to fix this inconsistency.

Out of the two approaches suggested, I am currently implementing the second one. The `rc_` parameter of the result is right now being set as the `current payload size` (i.e. the size of the packet passed into writePacket() ). Under this scenario:

1. We will return the current payload_len as `rc_` whenever the result is `WRITE_STATUS_OK` or  `WRITE_STATUS_DATA_BUFFERED` since these results allow the caller to drain the buffers.
2. In all other cases the `rc_` will be returned as ZERO.

NOTE: For external flush, the result will always return `rc_` as zero. Since effectively the current payload_len is ZERO for external flush.